roaring/roaring.go

```diff
@@ -1563,6 +1563,127 @@ func differenceBitmapBitmap(a, b *container) *container {
 	}
 	return output
 }
+func (b *Bitmap) Xor(other *Bitmap) *Bitmap {
+	output := &Bitmap{}
+
+	ki, ci := b.keys, b.containers
+	kj, cj := other.keys, other.containers
+
+	for {
+		var key uint64
+		var container *container
+
+		ni, nj := len(ki), len(kj)
+		if ni == 0 && nj == 0 { // eof(i,j)
+			break
+		} else if ni == 0 || (nj != 0 && ki[0] > kj[0]) { // eof(i) or i > j
+			key, container = kj[0], cj[0].clone()
+			kj, cj = kj[1:], cj[1:]
+		} else if nj == 0 || (ki[0] < kj[0]) { // eof(j) or i < j
+			key, container = ki[0], ci[0].clone()
+			ki, ci = ki[1:], ci[1:]
+		} else { // i == j
+			key, container = ki[0], xor(ci[0], cj[0])
+			ki, ci = ki[1:], ci[1:]
+			kj, cj = kj[1:], cj[1:]
+		}
+
+		output.keys = append(output.keys, key)
+		output.containers = append(output.containers, container)
+	}
+
+	return output
+}
+
+func xor(a, b *container) *container {
+	if a.isArray() {
+		if b.isArray() {
+			return xorArrayArray(a, b)
+		} else {
+			return xorArrayBitmap(a, b)
+		}
+	} else {
+		if b.isArray() {
+			return xorArrayBitmap(b, a)
+		} else {
+			return xorBitmapBitmap(a, b)
+		}
+	}
+}
+
+func xorArrayArray(a, b *container) *container {
+	output := &container{}
+	na, nb := len(a.array), len(b.array)
+	for i, j := 0, 0; ; {
+		if i >= na && j >= nb {
+			break
+		} else if i < na && j >= nb {
+			output.add(a.array[i])
+			i++
+			continue
+		} else if i >= na && j < nb {
+			output.add(b.array[j])
+			j++
+			continue
+		}
+
+		va, vb := a.array[i], b.array[j]
+		if va < vb {
+			output.add(va)
+			i++
+		} else if va > vb {
+			output.add(vb)
+			j++
+		} else { //==
+			i++
+			j++
+		}
+	}
+	return output
+}
+
+func xorArrayBitmap(a, b *container) *container {
+	output := &container{}
+	itr := newBufIterator(newBitmapIterator(b.bitmap))
+	for i := 0; ; {
+		vb, eof := itr.next()
+		if i >= len(a.array) && eof {
+			break
+		} else if i >= len(a.array) {
+			output.add(vb)
```

Comment:
The cardinality will most likely be significantly different right? Since array type containers are guaranteed to be < 4096 bits, and bitmap containers will be 4096-65535 bits. 

My thinking is that because the bitmap will usually have many more bits set and they will probably all appear in  the output (with the exception of the few which happen to collide with the array), it makes sense to copy the bitmap and set or unset the comparatively few bits which are in the array.

We should certainly benchmark - especially since we already have this implementation, but I think my way is better 😛.

Also I'm just trying to slow you down so that I can get RLE in first 😈